File: PhenixServer/communication/udpserver.py
#!/usr/bin/python
import socket, sys
SERVER_PORT = 1234

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(('', SERVER_PORT))

user_list = []
class User:
    def __init__(self,name,address):
        self.user_name = name
        self.public_address = address
    def __str__(self):
        return "[%s,%s]" % (self.user_name , self.public_address)

def dictFromString(headder):
	line_list = headder.split('\n')
	ret = {}
	for one_line in line_list:
		compoments = one_line.split(': ')
		if len(compoments) >= 2 and len(compoments[0]) > 0:
			ret[compoments[0]] = compoments[1]
	return ret

def server_handle():
	while True:
		data, address = sock.recvfrom(8192)
		if data == None or len(data) == 0:
			continue
		command = dictFromString(data)
		command_type = command['type']
		if command_type == 'login':
			name = command['user_name']
			address_str = command['address']
			user_private_ip = address_str.split(':')[0]
			user_private_port = long(address_str.split(':')[1])
			# now_user = User(name, address, (user_private_ip, user_private_port))
			now_user = None
			for one_user in user_list:
				# if one_user.user_name == name and one_user.public_address == address and one_user.private_address == (user_private_ip, user_private_port):
				if one_user.user_name == name and one_user.public_address == address:
					now_user = one_user
					break
			if now_user == None:
				print "add user"
				now_user = User(name, address)
				user_list.append(now_user)
				send_add_user(now_user)
				
		elif command_type == 'logout':
			name = command['user_name']
			address_str = command['address']
			user_private_ip = address_str.split(':')[0]
			user_private_port = long(address_str.split(':')[1])
			now_user = None
			for one_user in user_list:
				# if one_user.user_name == name and one_user.public_address == address and one_user.private_address == (user_private_ip, user_private_port):
				if one_user.user_name == name and one_user.public_address == address:
					now_user = one_user
					break
			if now_user != None and len(user_list) != 0:
				user_list.remove(now_user)
				send_remove_user(now_user)
			
		elif command_type == 'get_all_user':
			for one_user in user_list:
				if one_user.user_name != None and one_user.public_address != None:
					send_add_user(one_user)
		
		elif command_type == 'p2ptrans':
			print data
			connect_address_str = command['connect_address']
			connect_ip = connect_address_str.split(':')[0]
			connect_port = long(connect_address_str.split(':')[1])
			p2pconnect((connect_ip, connect_port), address)
	
		elif command_type == 'puserlist':
			print_user_list()
		elif command_type == 'echo':
			print "echo: "
			print data
			print address
			sendnum = sock.sendto(data, address)
		
def send_remove_user(user):
	headder = "type: rmuser\nuser_name: %s\naddress: %s:%d\n" % (user.user_name, user.public_address[0] , user.public_address[1])
	for user in user_list:
		sock.sendto(headder, user.public_address)
			
def send_add_user(user):
	headder = "type: adduser\nuser_name: %s\naddress: %s:%d\n" % (user.user_name, user.public_address[0] , user.public_address[1])
	for user in user_list:
		sock.sendto(headder, user.public_address)
			
def p2pconnect(connect_addr, address):
	headder = "type: p2pconnect\naddress: %s:%d\n" % address
	sock.sendto(headder, connect_addr)
	
def print_user_list():
	for user in user_list:
		print user
if __name__ == '__main__':
	server_handle()
	

File: PhenixServer/communication/udpserver.cpp
/* P2P 程序服务端
 * 
 * 文件名：P2PServer.c
 *
 * 日期：2004-5-21
 *
 * 作者：shootingstars(zhouhuis22@sina.com)
 *
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pthread.h>

#include "protos.h"
#include "Exceptions.h"

#define SOCKET int

UserList ClientList;

void InitWinSock()
{
  int thisSocket;
	struct sockaddr_in destination;

	destination.sin_family = AF_INET;
	thisSocket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP); 
	if (thisSocket < 0)
	{
  	printf("Error");
    exit(1);
	}
	printf("Startup\n");
}


SOCKET mksock(int type)
{
	SOCKET sock = socket(AF_INET, type, 0);
	if (sock < 0)
	{
        printf("create socket error");
		exit(1);
	}
	return sock;
}

stUserListNode GetUser(char *username)
{
	for(UserList::iterator UserIterator=ClientList.begin();
						UserIterator!=ClientList.end();
							++UserIterator)
	{
		if( strcmp( ((*UserIterator)->userName), username) == 0 )
			return *(*UserIterator);
	}
  printf("not find this user");
  exit(1);
}

int main(int argc, char* argv[])
{
	try{
		InitWinSock();
		
		SOCKET PrimaryUDP;
		PrimaryUDP = mksock(SOCK_DGRAM);

		sockaddr_in local;
		local.sin_family=AF_INET;
		local.sin_port= htons(SERVER_PORT); 
		local.sin_addr.s_addr = htonl(INADDR_ANY);
		int nResult=bind(PrimaryUDP,(sockaddr*)&local,sizeof(sockaddr));
		if(nResult)
    {
      printf("bind error");
      exit(1);
    }
			

		sockaddr_in sender;
		stMessage recvbuf;
		memset(&recvbuf,0,sizeof(stMessage));

		// 开始主循环.
		// 主循环负责下面几件事情:
		// 一:读取客户端登陆和登出消息,记录客户列表
		// 二:转发客户p2p请求
		for(;;)
		{
			int dwSender = sizeof(sender);
			int ret = recv(PrimaryUDP, (char *)&recvbuf, sizeof(stMessage), 0);
			if(ret <= 0)
			{
				printf("recv error");
				continue;
			}
			else
			{
				int messageType = recvbuf.iMessageType;
				switch(messageType){
				case LOGIN:
					{
						//  将这个用户的信息记录到用户列表中
						printf("has a user login : %s\n", recvbuf.message.loginmember.userName);
						stUserListNode *currentuser = new stUserListNode();
						strcpy(currentuser->userName, recvbuf.message.loginmember.userName);
						currentuser->ip = ntohl(sender.sin_addr.s_addr);
						currentuser->port = ntohs(sender.sin_port);
						
						ClientList.push_back(currentuser);

						// 发送已经登陆的客户信息
						int nodecount = (int)ClientList.size();
						send(PrimaryUDP, (const char*)&nodecount, sizeof(int), 0);
						for(UserList::iterator UserIterator=ClientList.begin();
								UserIterator!=ClientList.end();
								++UserIterator)
						{
							send(PrimaryUDP, (const char*)(*UserIterator), sizeof(stUserListNode), 0); 
						}

						break;
					}
				case LOGOUT:
					{
						// 将此客户信息删除
						printf("has a user logout : %s\n", recvbuf.message.logoutmember.userName);
            // UserList::iterator removeiterator = false;
            // for(UserList::iterator UserIterator=ClientList.begin();
            //   UserIterator!=ClientList.end();
            //   ++UserIterator)
            // {
            //   if( strcmp( ((*UserIterator)->userName), recvbuf.message.logoutmember.userName) == 0 )
            //   {
            //     removeiterator = UserIterator;
            //     break;
            //   }
            // }
            // if(removeiterator != NULL)
            //   ClientList.remove(*removeiterator);
						break;
					}
				case P2PTRANS:
					{
						// 某个客户希望服务端向另外一个客户发送一个打洞消息
						printf("%s wants to p2p %s\n",inet_ntoa(sender.sin_addr),recvbuf.message.translatemessage.userName);
						stUserListNode node = GetUser(recvbuf.message.translatemessage.userName);
						sockaddr_in remote;
						remote.sin_family=AF_INET;
						remote.sin_port= htons(node.port); 
						remote.sin_addr.s_addr = htonl(node.ip);

						in_addr tmp;
						tmp.s_addr = htonl(node.ip);
						printf("the address is %s,and port is %d\n",inet_ntoa(tmp), node.port);

						stP2PMessage transMessage;
						transMessage.iMessageType = P2PSOMEONEWANTTOCALLYOU;
						transMessage.iStringLen = ntohl(sender.sin_addr.s_addr);
						transMessage.Port = ntohs(sender.sin_port);
                        
						send(PrimaryUDP,(const char*)&transMessage, sizeof(transMessage), 0);

						break;
					}
				
				case GETALLUSER:
					{
						int command = GETALLUSER;
						send(PrimaryUDP, (const char*)&command, sizeof(int), 0);

						int nodecount = (int)ClientList.size();
						send(PrimaryUDP, (const char*)&nodecount, sizeof(int), 0);

						for(UserList::iterator UserIterator=ClientList.begin();
								UserIterator!=ClientList.end();
								++UserIterator)
						{
							send(PrimaryUDP, (const char*)(*UserIterator), sizeof(stUserListNode), 0); 
						}
						break;
					}
				}
			}
		}

	}
	catch(Exception &e)
	{
		printf(e.GetMessage());
		return 1;
	}

	return 0;
}

